### ISSUE TITLE
Tests for rules and the whole pipeline

### ISSUE THREAD

**Issue by CholoTook** [snakemake/snakemake#599]
**Is your feature request related to a problem? Please describe.**
We want to create a test for the whole pipeline that can be run on every commit.

**Describe the solution you'd like**
I'd like a special flag to call on the command line that runs the pipeline in 'test mode', such that we pick up test inputs and test background data. Some rules would switch into 'mock' mode (e.g. when the rule /should/ download a large data set and we can assume that it will work).

**Describe alternatives you've considered**
I've been googling around for a solution. I guess we create some 'test' target that we can fire off using a commit hook? Perhaps all I need is a tutorial on how to do that. I was surprised I didn't find 'testing snakemake, best practices' document.

**Additional context**
This is for a CI/CD environment.

**Comment by CholoTook** [comment]
BTW, I asked a question here: https://bioinformatics.stackexchange.com/questions/14303/how-do-i-write-tests-for-a-snakemake-pipeline

**Comment by nhartwic** [comment]
For testing your entire workflow, its probably best to simply have a minimal example dataset to run on. I think the better question is if it makes sense to try to test your entire workflow in that way. Given the way snakemake works, with DAG building at execution time, potentially in different ways from pipeline run to pipeline run, unit tests of individual rules + dry runs probably make more sense.

There isn't really any built in support for rule level unit tests that I'm aware of. It could probably be added without too much trouble though supporting all the potential back ends might be a pain. Basically, each rule needs 3 new fields: test\_input, test\_validator, test\_wildcards...

1. test\_input : A list/dict of file/url paths pointing to input files the rule needs to execute in test mode
2. test\_wildcards : a dict mapping from wildcards to values to populate the output with. Defaults to empty dict.
3. test\_validator : A function that will run after the rule and should return True if the rule passes the test and False if the rule failed the test. Defaults to 'lambda output : True '

...This kind of design leaves the bulk of the actual testing to the person designing the workflow. All snakemake can do is test if the output files were generated. Any further testing of the output files should be done by the validator function. Its also possible that each of these should permit accepting a list of the defined values so as to permit multiple unit tests per rule. Thoughts?

**Comment by CholoTook** [comment]
I agree that the main thing is to 1) unit test the individual rules, and 2) dry run the pipeline in various ways. However, testing the pipeline from 'end to end' also feels like an important requirement.

The new fields sound reasonable, but I'd have to think about / see / try using it in anger to see if it 'feels' right too. I've only ever skimmed the surface of unit testing and testing frameworks in Python (e.g. I'm a noob). Somehow I assume that it should 'feel' the same in snakemake... e.g. I was thinking something like...

```
rule align:
    input:
        data/reference.fa,
        data/sample.fastq
    output:
        data/sample.bam
    shell:
        aligner {input} -o {output}
    conda:
        ...
    ...

rule align_test
    input:
        test/data/reference_test.fa,
        test/data/sample_test.fa
    output:
        test/data/sample_test.bam
    test_shell:
        if [ $(md5sum test/data/sample_test.bam)== deadbeef ]; then echo true; else fail; fi
```

which is ugly, but you get the idea I guess.

I imagine you could have `test_shell` or `test_script` or whatever. All you do in the '_test' version of the rule is override the input and output and add the test_ field, which should either pass or fail just like the corresponding field in the rule.

That way you keep your rules clean of testing 'junk' and you can write your tests relatively independently of the rules and without writing 'dedicated' validator functions (if you don't want to). Each test just inherits missing fields (and possibly wildcards?) from the corresponding rule.

Just another idea.

As I said, I'm quite surprised there isn't a 'best practice' for testing rules already, so I think it makes sense to take the time to do this as simply / easily / cleanly as possible.

Cheers,

**Comment by cademirch** [comment]
See https://snakemake.readthedocs.io/en/stable/snakefiles/testing.html
